keadm/cmd/keadm/app/cmd/beta/config.go

```diff
@@ -219,14 +228,23 @@ func GetPauseImage() string {
 }
 
 // dockerPullImage uses Docker to pull the image
-func dockerPullImage(image string) error {
-	var err error
-	for i := 0; i < PullImageRetry; i++ {
-		pullCommand := fmt.Sprintf("docker pull %s", image)
-		err = util.NewCommand(pullCommand).Exec()
-		if err == nil {
-			return nil
-		}
+func dockerPullImage(ctx context.Context, image string, cli *dockerclient.Client) error {
+	args := filters.NewArgs()
+	args.Add("reference", image)
+	list, err := cli.ImageList(ctx, dockertypes.ImageListOptions{Filters: args})
+	if err != nil {
+		return err
+	}
+	if len(list) > 0 {
+		return nil
+	}
+
+	rc, err := cli.ImagePull(ctx, image, dockertypes.ImagePullOptions{})
+	if err != nil {
+		return err
+	}
+	if _, err = io.Copy(io.Discard, rc); err != nil {
+		return err
 	}
-	return fmt.Errorf("dcoker pull image failed: %v", err)
+	return rc.Close()
```

Comment:
It might be better to use `defer` here

Reply:
it's in the last line of the function, so it seems duplicate to use defer to close the io :)

Reply:
```golang
if _, err = io.Copy(io.Discard, rc); err != nil {
	return err
}
```
Here if err is returned, we should also call the close method

Reply:
fixed, please look again :)